compiler/src/iree/compiler/Codegen/Common/TransformExtensions/CommonExtensions.cpp

```diff
@@ -1021,10 +1021,107 @@ static DiagnosedSilenceableFailure testMatchCallbackCallback(
   return DiagnosedSilenceableFailure::success();
 }
 
+/// Match callback for a reduction with optional leading and trailing
+/// elementwise operations. Matches *the first* occurrence of such a reduction
+/// within an op associated with the given handle.
+///
+/// Input handles:
+///
+///   - container op, must be associated with one operation.
+///
+/// Output handles:
+///
+///   - leading elementwise op, if any;
+///   - the "fill" op preceding the reduction;
+///   - reduction op;
+///   - trailing elementwise op, if any.
+static DiagnosedSilenceableFailure reductionCallback(
+    transform_dialect::MatchCallbackResult &res, Location loc,
+    const transform::TransformState &state, ValueRange handles) {
+  if (handles.size() != 1 || state.getPayloadOps(handles[0]).size() != 1) {
+    return emitSilenceableFailure(loc)
+           << "expected one handle to one operation";
+  }
+
+  transform_dialect::StructuredOpMatcher pattern, fill, leadingEltwise,
+      trailingEltwise;
+  makeGPUReductionMatcher(pattern, fill, leadingEltwise, trailingEltwise);
+
+  // TODO: need a mechanism for this to go around the entire IR,
```

Comment:
Interesting, so you see this evolving into a grassroots full graph matcher over time? :)

This is significantly more than IREE needs/can make use of right now because we don't control the dispatch region formation but it is great to start to think about.

Reply:
Yes, the full integration looks like finding relevant subgraphs in a bigger graph.